src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,230 @@
+            Node<T> newNode = new Node<>(predecessor, e, null);
+            if (predecessor == null) {
+                firstNode = newNode;
+            } else {
+                predecessor.next = newNode;
+            }
+            predecessor = newNode;
+        }
+
+        if (successor == null) {
+            lastNode = predecessor;
+        } else {
+            predecessor.next = successor;
+            successor.previous = predecessor;
+        }
+        size += numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkedElementIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkedElementIndex(index);
+        Node<T> x = node(index);
+        T oldValue = x.item;
+        x.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkedElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = firstNode; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = firstNode; x != null; x = x.next) {
+                if (object.equals(x.item)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> x) {
+
+        final T element = x.item;
+        final Node<T> next = x.next;
+        final Node<T> previous = x.previous;
+
+        if (previous == null) {
+            firstNode = next;
+        } else {
+            previous.next = next;
+            x.previous = null;
+        }
+
+        if (next == null) {
+            lastNode = previous;
+        } else {
+            next.previous = previous;
+            x.next = null;
+        }
+
+        x.item = null;
+        size--;
+        return element;
+    }
+
+    private void checkedElementIndex(int index) {
+        if (!isElementIndex(index)) {
+            throw new IndexOutOfBoundsException(index + " out of bounds");
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException(index);
+        }
+    }
+
+    private boolean isElementIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    private Node<T> node(int index) {
```

Comment:
Don't iterate your nodes from the head to the tail or vice versa. Optimize `node(int index)` method to iterate only the first or second half of the list, depending on the index value.